lib/Infer/InstSynthesis.cpp

```diff
@@ -0,0 +1,1373 @@
+        if ((In.first == 0 || In.second == 0) && !isWiringInvalid(In, OpLoc)) {
+          Match = In;
+          if (DebugLevel > 3)
+            llvm::outs() << getLocVarStr(Match);
+          FoundMatch = true;
+          break;
+        }
+      }
+      if (DebugLevel > 3)
+        llvm::outs() << "\n";
+    }
+    if (FoundMatch)
+      break;
+  }
+
+  return Match;
+}
+
+Inst *InstSynthesis::createJunkFreeInst(Inst::Kind Kind, unsigned Width,
+                                        std::vector<Inst *> &Ops,
+                                        InstContext &IC) {
+  switch (Kind) {
+  case Inst::Add:
+  case Inst::AddNSW:
+  case Inst::AddNUW:
+  case Inst::AddNW: {
+    if (Ops[0] == IC.getConst(APInt(Width, 0)))
+      return Ops[1];
+    else if (Ops[1] == IC.getConst(APInt(Width, 0)))
+      return Ops[0];
+    break;
+  }
+
+  case Inst::Sub:
+  case Inst::SubNSW:
+  case Inst::SubNUW:
+  case Inst::SubNW: {
+    if (Ops[0] == Ops[1])
+      return IC.getConst(APInt(Width, 0));
+    else if (Ops[1] == IC.getConst(APInt(Width, 0)))
+      return Ops[0];
+    break;
+  }
+
+  case Inst::Mul:
+  case Inst::MulNSW:
+  case Inst::MulNUW:
+  case Inst::MulNW: {
+    if (Ops[0] == IC.getConst(APInt(Width, 1)))
+      return Ops[1];
+    else if (Ops[1] == IC.getConst(APInt(Width, 1)))
+      return Ops[0];
+    break;
+  }
+
+  case Inst::UDiv:
+  case Inst::SDiv:
+  case Inst::UDivExact:
+  case Inst::SDivExact: {
+    if (Ops[1] == IC.getConst(APInt(Width, 1)))
+      return Ops[0];
+    break;
+  }
+
+  case Inst::And:
+  case Inst::Or: {
+    if (Ops[0] == Ops[1])
+      return Ops[0];
+    break;
+  }
+
+  case Inst::Xor: {
+    if (Ops[0] == Ops[1])
+      return IC.getConst(APInt(Width, 0));
+    break;
+  }
+
+  case Inst::Shl:
+  case Inst::ShlNSW:
+  case Inst::ShlNUW:
+  case Inst::ShlNW:
+  case Inst::LShr:
+  case Inst::LShrExact:
+  case Inst::AShr:
+  case Inst::AShrExact: {
+    if (Ops[1] == IC.getConst(APInt(Width, 0)))
+      return Ops[0];
+    break;
+  }
+
+  case Inst::Select: {
+    if (Ops[1] == Ops[2])
+      return Ops[1];
+    break;
+  }
+
+  case Inst::ZExt:
+  case Inst::SExt:
+  case Inst::Trunc: {
+    if (Width == Ops[0]->Width)
+      return Ops[0];
+    if (Ops[0]->K == Inst::Const)
+      return IC.getConst(APInt(Width, Ops[0]->Val.getZExtValue()));
+    break;
+  }
+
+  case Inst::Eq: {
+    if (Ops[0] == Ops[1])
+      return IC.getConst(APInt(1, true));
+    break;
+  }
+
+  case Inst::Ne: {
+    if (Ops[0] == Ops[1])
+      return IC.getConst(APInt(1, false));
+    break;
+  }
+
+  case Inst::Ult:
+  case Inst::Slt: {
+    if (Ops[0] == Ops[1])
+      return IC.getConst(APInt(1, false));
+    break;
+  }
+
+  case Inst::CtPop:
+  case Inst::BSwap: {
+    if (Ops[0] == IC.getConst(APInt(Width, 0)))
+      return IC.getConst(APInt(Width, 0));
+    break;
+  }
+
+  case Inst::Cttz:
+  case Inst::Ctlz: {
+    if (Ops[0] == IC.getConst(APInt(Width, 0)))
+      return IC.getConst(APInt(Width, Width));
```

Comment:
It's a nit, but do we want to use curlies around a case that doesn't declare any local variables?